public/app/features/explore/TraceView/createSpanLink.tsx

```diff
@@ -383,10 +383,10 @@ function getQueryForLoki(
 
   let expr = '{${__tags}}';
   if (filterByTraceID && span.traceID) {
-    expr += ' |="${__span.traceId}"';
+    expr += ' | trace_id="${__span.traceId}"';
```

Comment:
Doesn't this new syntax only work when logs are persisted with the Loki format for Otel logs relying on Loki log record metadata and labels?

Reply:
I haven't tested this with anything other than the latest version of Loki and the latest version of Grafana. My assumption would be that this should be "the one that works".

Should Grafana default to working with the old format, or the new format? Or should there be a more complex solution that changes the expression based on the data source? Does the data source expose enough metadata for such a solution to leverage? Should it work with the new format while awaiting a more complex solution for all/more cases?

I can instead/additionally add a docs PR that explains all this, because currently it relies on finding the ticket that describes the problem.

Reply:
@stevendungan @sandeepsukhani @vlad-diachenko Please have a look, @JamesBelchamber summarized very well the questions

Reply:
Hey @JamesBelchamber , @cyrille-leclerc . 
This change is dangerous because it changes the behavior of the mapping for everyone. 
Even if it works for OTEL data, because the `traceId` and `spanId` fields from the payload are attached to the logs as structured metadata labels `trace_id` and `span_id` accordingly, the solution won't work for the rest of the users or rest of the data that is ingested in regular Loki format. Moreover, using regular Loki push endpoint, the users can use different label names for trace and span IDs : `spanID` ,`span-id`,`spanId`, `traceID`, `trace-id`, `traceId`, etc... 

Also, taking into account that we already can modify Loki query by toggling the flag: "Use custom query", I would vote to avoid merging this change... 
![image](https://github.com/user-attachments/assets/4c0626f3-dac6-480d-b6fa-ec7aff506300)


At the same time, I do understand that the current solution when Grafana puts trace id and span id as line filters (`{...} |= "sometraceid" |= "somespanid"`) does not cover all possible cases, because the logline itself can do not contain the trace id and span id, but they are attached as structured metadata only. 

I would vote for improving the UI/UX by making the configuration transparent and clear for the user. From my point of view, we need to explicitly show how the query is constructed when the user enables flags `Filter by trace ID` or `Filter by span ID`. when the state of the toggle is changed,  UI can update the field with the query, so the user immediately sees where Grafana puts traceId and spanId in the query, and in case he wants to make the query more precise for his data, he can immediately modify the query.

Reply:
I don't really mind how this is handled, and I'm able to put in the custom query for our own use.

It's frustrating that it doesn't work with the latest Loki and Tempo, with spans and logs coming from an OpenTelemetry Collector. I would (without data to back it up) assume this would be "the most common approach" for users looking to leverage this feature, with adding the span/trace IDs to unstructured logs being more esoteric (again, entirely personal perspective).

Agreed that a more complex solution would be better, e.g. having a toggle to set whether the trace/span id is in the structured or unstructured part of the log line (and then setting a default label when it's set to structured). It would be nice if, by default, it "just worked" for Traces and Logs coming out of the OpenTelemetry Collector.

If you wish to close off this PR then feel free to go ahead :)

Reply:
Hey @JamesBelchamber. 
Yes, it definitely should work by default regardless of where the trace ID is defined, whether in the log line or structured metadata. 
We are already working on a a permanent fix that would satisfy the users' needs who ingest data through Loki push endpoint and/or Loki native OTLP endpoint, because the different push endpoints can be used within a single company. 
The default query to search by trace ID will look like 
```
{${__tags}} | label_format log_line_contains_trace_id=`{{ contains "${__span.traceId}" __line__  }}` 
| log_line_contains_trace_id="true" OR trace_id="${__span.traceId}"
```

So, I am closing this PR in favor of the solution we implement shortly. I will link that PR once it's opened.